File: t_2/src/main/java/com/gmail/t_2/MainActivity.java
package com.gmail.t_2;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.gmail.t_2.service.MathService;
import com.gmail.t_2.service.impl.MathServiceImpl;

import java.util.ArrayList;
import java.util.List;

import static com.gmail.t_2.SecondActivity.RESULT_1;
import static com.gmail.t_2.SecondActivity.RESULT_2;
import static com.gmail.t_2.SecondActivity.RESULT_3;

@SuppressLint("SetTextI18n")
public class MainActivity extends AppCompatActivity {


    public static final String VALUE_LIST = "valueList";
    private MathService mathService = new MathServiceImpl();

    private List<Integer> valueList;
    private TextView avgText;
    private TextView sumText;
    private TextView quotientText;
    private TextView text_1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        text_1 = findViewById(R.id.text_1);

        avgText = findViewById(R.id.result_1);
        sumText = findViewById(R.id.result_2);
        quotientText = findViewById(R.id.result_3);
        avgText.setText(R.string.average);
        sumText.setText(R.string.sum);
        quotientText.setText(R.string.quotient);

        valueList = mathService.generateNewList();
        final StringBuilder values = getStringValues();
        text_1.setText("Generated numbers :  " + values.toString());

        findViewById(R.id.btn).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(MainActivity.this, SecondActivity.class);

                intent.putIntegerArrayListExtra(VALUE_LIST, (ArrayList<Integer>) valueList);
                startActivityForResult(intent, 1);
            }
        });

        findViewById(R.id.btnRestart).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(MainActivity.this, ThirdActivity.class);
                startActivityForResult(intent, 666);
            }
        });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (requestCode == 1 && resultCode == Activity.RESULT_OK && data != null) {
            Double average = data.getDoubleExtra(RESULT_1, 0.0);
            Double sum = data.getDoubleExtra(RESULT_2, 0.0);
            Double quotient = data.getDoubleExtra(RESULT_3, 0.0);

            avgText.setText("Average of numbers:  " + average);
            sumText.setText("Sum of numbers:  " + sum);
            quotientText.setText("Quotient of numbers:  " + quotient);
        }
        if (requestCode == 666 && resultCode == Activity.RESULT_OK) {
            valueList = mathService.generateNewList();
            avgText.setText(R.string.average);
            sumText.setText(R.string.sum);
            quotientText.setText(R.string.quotient);

            final StringBuilder values = getStringValues();
            text_1.setText("Generated numbers :  " + values.toString());
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    private StringBuilder getStringValues() {
        final StringBuilder values = new StringBuilder();
        for (Integer num : valueList) {
            values.append(num).append("; ");
        }
        return values;
    }

}

File: task_8_weather_mvp/src/main/java/com/gmail/task_8_weather_mvp/MainActivity.kt
package com.gmail.task_8_weather_mvp

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.gmail.task_8_weather_mvp.adapter.WeatherAdapter
import com.gmail.task_8_weather_mvp.presenter.HourlyPresenterImpl
import com.gmail.task_8_weather_mvp.presenter.viewmodel.WeatherViewModel
import com.gmail.task_8_weather_mvp.view.HourlyWeatherView
import com.google.android.material.floatingactionbutton.FloatingActionButton

class MainActivity : AppCompatActivity(), HourlyWeatherView {

    private lateinit var hourlyPresenter: HourlyPresenterImpl
    private val weatherAdapter by lazy { WeatherAdapter() }
    private lateinit var recyclerView: RecyclerView
    private lateinit var cityName: TextView
    private lateinit var weatherInfo: TextView
    private lateinit var temperatureText: TextView
    private lateinit var settingsButton: ImageButton
    private lateinit var weatherIcon: ImageView
    private lateinit var selectCityButton: FloatingActionButton
    private lateinit var units: String

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        units = loadUnitStatus()
        hourlyPresenter = HourlyPresenterImpl(this, applicationContext as App)
        initViews()
        settingsButton.setOnClickListener {
            val intent = Intent(this, SettingsActivity::class.java)
            startActivityForResult(intent, 1)
        }
        selectCityButton.setOnClickListener {
            val intent = Intent(this, SelectCityActivity::class.java)
            startActivityForResult(intent, 2)
        }
        recyclerView.apply {
            adapter = weatherAdapter
            layoutManager = LinearLayoutManager(this@MainActivity, RecyclerView.VERTICAL, false)
        }
        hourlyPresenter.getHourlyWeather()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == 1 && resultCode == Activity.RESULT_OK && data != null) {
            units = data.getStringExtra(UNITS).toString()
            hourlyPresenter.getHourlyWeather()
        } else if (requestCode == 2 && resultCode == Activity.RESULT_OK) {
            hourlyPresenter.getHourlyWeather()
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    private fun initViews() {
        recyclerView = findViewById(R.id.hourlyWeatherRecyclerView)
        temperatureText = findViewById(R.id.temperatureText)
        weatherInfo = findViewById(R.id.weatherInfo)
        settingsButton = findViewById(R.id.settingsButton)
        selectCityButton = findViewById(R.id.selectCityButton)
        cityName = findViewById(R.id.cityName)
        weatherIcon = findViewById(R.id.weatherIcon)
    }

    @SuppressLint("SetTextI18n")
    override fun fillPage(weatherViewModel: WeatherViewModel) {
        weatherAdapter.putAllWeatherList(weatherViewModel.hourly)
        if (loadUnitStatus() == METRIC) {
            temperatureText.text = weatherViewModel.currentTemp.toString() + "C"
        } else {
            temperatureText.text = weatherViewModel.currentTemp.toString() + "K"
        }
        if(weatherViewModel.currentTemp > 0 || weatherViewModel.currentTemp > 273.15){
            weatherIcon.setImageResource(R.drawable.ic_sunny)
        } else {
            weatherIcon.setImageResource(R.drawable.ic_cold)
        }
        weatherInfo.text = weatherViewModel.currentMainWeather
        cityName.text = weatherViewModel.currentCityName
    }

    private fun loadUnitStatus(): String {
        val sharedPrefs = applicationContext.getSharedPreferences(UNITS, Context.MODE_PRIVATE)
        return sharedPrefs.getString(UNITS, METRIC).toString()
    }
}